### PR TITLE
Implement custom mutability error for initialized simulations

### LINKED ISSUE #1043: Add new error class for attributes that are no longer settable after attaching to a simulation
## Description

An custom error class that signified an attribute cannot be set after a simulation has been run, and to set, the object must be removed from the simulation would be helpful to users.

## Proposed solution
Create a new error class and modify `ParameterDict` to use it.

## Developer

Anyone who wants to can. The changes are minimal. I may if given the time in the future.

### PR DESCRIPTION
**Pull request by AlainKadar**
## Description

Define a custom error which explains that read only attributes cannot be modified after a simulation has been run. Modify attribute setting methods so that this error is called when appropriate

## Motivation and context

Resolves #1043 

## How has this been tested?

Attempting to change the filename of a gsd writer successfully raises the error

## Change log

```

```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/operation.py [hunk 1]
```diff
@@ -81,8 +82,7 @@ def _setattr_param(self, attr, value):
                 setattr(self._cpp_obj, attr, new_value)
             except (AttributeError):
                 self._param_dict[attr] = old_value
-                raise AttributeError("{} cannot be set after cpp"
-                                     " initialization".format(attr))
+                raise MutabilityError
```
Comment (b-butler) [inline hoomd/operation.py:85]:
```suggestion
                raise MutabilityError(attr)
```

**Review by b-butler — CHANGES_REQUESTED** [review]
Thanks for the PR. I have one suggestion, and don't forget to update the tests for classes that should raise this new error.

File: hoomd/pytest/test_box_resize.py [hunk 2]
```diff
@@ -233,3 +234,19 @@ def test_pickling(simulation_factory, two_particle_snapshot_factory,
...
+
+
+def test_mutability_error(simulation_factory, two_particle_snapshot_factory):
+    filt = hoomd.filter.All()
+    sim = simulation_factory(two_particle_snapshot_factory())
+    trig = hoomd.trigger.Periodic(1)
+    box1 = hoomd.Box.cube(L=10)
+    box2 = hoomd.Box.cube(L=12)
+    var = hoomd.variant.Ramp(10, 12, 0, 100)
+    box_op = hoomd.update.BoxResize(box1, box2, var, trig)
```
Comment (joaander) [inline hoomd/pytest/test_box_resize.py:246]:
```suggestion
    box_op = hoomd.update.BoxResize(box1=box1, box2=box2, variant=var, trigger=trig)
```
The `master` branch changed the order of these arguments, causing the test to fail. Use keyword arguments so they are passed without ambiguity.

**Comment by joaander** [comment]
To help move this forward, I committed the suggested changes and fixed a test that was failing because of changes on merged in #1115.

**Comment by joaander** [comment]
Suggested change log entry:
```
Added:

- ``MutabilityError`` exception when setting read-only operation parameters.
```

Please suggest change log entries in future pull request descriptions.

**Comment by AlainKadar** [comment]
Thanks for commits and fixes, @joaander

**Comment by joaander** [comment]
This pull request is passing and ready for final review.

**Review by b-butler — APPROVED** [review]
LGTM

**Pull request merged**